### PR TITLE
Improving Unit Test Coverage

### PR DESCRIPTION
**Pull request by hiranya911**
This PR adds more unit tests to cover classes like `Connection`, `PersistentConnectionImpl` and `WebsocketConnection`. To enable this, I also made some minor modifications to the classes in question, so they are more amenable to unit testing. Overall this brings our unit test coverage to 80%.

### CODE REVIEW

File: src/test/java/com/google/firebase/database/core/RepoTest.java [hunk 1]
```diff
@@ -346,6 +387,82 @@ public void onComplete(DatabaseError error, boolean committed, DataSnapshot curr
     assertEquals(update, events.get(0).getValue());
   }
 
+  @Test
+  public void testTransactionError() throws InterruptedException {
+    final Repo repo = newRepo();
+    final List<DataSnapshot> events = new ArrayList<>();
+    final Path path = new Path("/txn_error");
```
Comment (schmidt-sebastian) [inline src/test/java/com/google/firebase/database/core/RepoTest.java]: I'm gonna use my knowledge that I just learned from Testing on the Toilet:

I think these tests would be easier to follow if you set up the mocks inline.

File: src/test/java/com/google/firebase/internal/GaeExecutorServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    assertEquals(4, threadFactory.counter.get());
+
+    executorService.shutdown();
+    assertTrue(executorService.awaitTermination(
+        TestUtils.TEST_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS));
+    assertTrue(executorService.isShutdown());
+    assertTrue(executorService.isTerminated());
+  }
+
+  private static class CountingThreadFactory implements ThreadFactory {
```
Comment (schmidt-sebastian) [inline src/test/java/com/google/firebase/internal/GaeExecutorServiceTest.java]: I was hoping this was part of the JDK :/ This seems incredible useful!

Comment (hiranya911) [reply src/test/java/com/google/firebase/internal/GaeExecutorServiceTest.java]: It seems to be supported by [`BasicThreadFactory`](https://commons.apache.org/proper/commons-lang/javadocs/api-3.0/org/apache/commons/lang3/concurrent/BasicThreadFactory.html#getThreadCount()) of commons-lang. But we are currently not depending on it.

File: src/test/java/com/google/firebase/cloud/StorageClientTest.java [hunk 3]
```diff
@@ -85,4 +90,40 @@ public void testAppDelete() throws IOException {
...
+
+  @Test
+  public void testNonExistingBucket() throws IOException {
+    FirebaseApp app = FirebaseApp.initializeApp(new FirebaseOptions.Builder()
+        .setCredentials(GoogleCredentials.fromStream(ServiceAccount.EDITOR.asStream()))
+        .setStorageBucket("mock-bucket-name")
+        .build());
+    Storage mockStorage = Mockito.mock(Storage.class);
+    StorageClient client = new StorageClient(app, mockStorage);
+    try {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/test/java/com/google/firebase/cloud/StorageClientTest.java]: `assertThrows` is a junit5 method. We are still on junit4. Lets plan to do upgrade in a future PR.

File: src/test/java/com/google/firebase/auth/FirebaseUserManagerTest.java [hunk 4]
```diff
@@ -153,6 +185,74 @@ public void testDeleteUser() throws Exception {
 
   @Test
   public void testGetUserHttpError() throws Exception {
+    UserManagerOp[] operations = new UserManagerOp[]{
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/test/java/com/google/firebase/auth/FirebaseUserManagerTest.java]: There are multiple operations that return the same error code.

File: src/test/java/com/google/firebase/auth/FirebaseUserManagerTest.java [hunk 5]
```diff
@@ -91,6 +91,22 @@ public void testGetUserByEmail() throws Exception {
...
+  public void testGetUserByEmailWithNotFoundError() throws Exception {
+    MockLowLevelHttpResponse response = new MockLowLevelHttpResponse();
+    response.setContent(TestUtils.loadResource("getUserError.json"));
+    MockHttpTransport transport = new MockHttpTransport.Builder()
+        .setLowLevelHttpResponse(response)
+        .build();
+    FirebaseUserManager userManager = new FirebaseUserManager(gson, transport, credentials);
+    try {
+      userManager.getUserByEmail("testuser@example.com");
+      fail("No error thrown for invalid response");
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/test/java/com/google/firebase/auth/FirebaseUserManagerTest.java]: We are still on junit4. Lets upgrade to junit5 in a separate PR.

File: src/main/java/com/google/firebase/database/core/Repo.java [hunk 6]
```diff
@@ -216,50 +215,7 @@ public void stopListening(QuerySpec query, Tag tag) {
...
-          operationLogger.debug("Restoring merge with id " + write.getWriteId());
-        }
-        connection.merge(write.getPath().asList(), write.getMerge().getValue(true), onComplete);
-        CompoundWrite resolved =
-            ServerValues.resolveDeferredValueMerge(write.getMerge(), serverValues);
-        serverSyncTree.applyUserMerge(
-            write.getPath(), write.getMerge(), resolved, write.getWriteId(), /*persist=*/ false);
-      }
-    }
+    // No-op
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/database/core/Repo.java]: Done

File: src/main/java/com/google/firebase/internal/GaeExecutorService.java [hunk 7]
```diff
@@ -157,9 +180,10 @@ public void execute(Runnable command) {
 
     private final ExecutorService executorService;
 
-    ExecutorWrapper(String threadName) {
-      GaeThreadFactory threadFactory = GaeThreadFactory.getInstance();
-      if (threadFactory.isUsingBackgroundThreads()) {
+    ExecutorWrapper(String threadName, ThreadFactory threadFactory) {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/internal/GaeExecutorService.java]: Done

File: src/main/java/com/google/firebase/database/connection/WebsocketConnection.java [hunk 8]
```diff
@@ -66,22 +65,23 @@
...
-
-  private WSClient createConnection(
-      HostInfo hostInfo, String optCachedHost, String optLastSessionId) {
-    String host = (optCachedHost != null) ? optCachedHost : hostInfo.getHost();
-    URI uri = HostInfo.getConnectionUrl(
-        host, hostInfo.isSecure(), hostInfo.getNamespace(), optLastSessionId);
-    return new NettyWebSocketClient(
-        uri, connectionContext.getUserAgent(), connectionContext.getThreadFactory(),
-        new WSClientHandlerImpl());
+    this.conn = clientFactory.newClient(
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/database/connection/WebsocketConnection.java]: We need to inject a reference to `this` into the created `WSClient`. Therefore, it must be done post construction.

File: src/main/java/com/google/firebase/database/connection/Connection.java [hunk 9]
```diff
@@ -51,12 +53,22 @@ public Connection(
...
+      Delegate delegate,
+      String optLastSessionId,
+      WebsocketConnectionFactory connFactory) {
     long connId = connectionIds++;
     this.hostInfo = hostInfo;
     this.delegate = delegate;
     this.logger = new LogWrapper(context.getLogger(), Connection.class, "conn_" + connId);
     this.state = State.REALTIME_CONNECTING;
-    this.conn = new WebsocketConnection(context, hostInfo, cachedHost, this, optLastSessionId);
+    this.conn = connFactory.newConnection(context, hostInfo, cachedHost, this, optLastSessionId);
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/database/connection/Connection.java]: We need to inject a reference to `this` into the created connection. Therefore, it must be done post construction.

File: src/main/java/com/google/firebase/database/connection/Connection.java [hunk 10]
```diff
@@ -262,6 +269,28 @@ public void injectConnectionFailure() {
     REALTIME_DISCONNECTED
   }
 
+  interface WebsocketConnectionFactory {
+    WebsocketConnection newConnection(
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/database/connection/Connection.java]: Done

File: src/main/java/com/google/firebase/database/connection/Connection.java [hunk 11]
```diff
@@ -51,12 +53,22 @@ public Connection(
...
       String optLastSessionId) {
+    this(context, hostInfo, cachedHost, delegate, optLastSessionId, new DefaultConnectionFactory());
+  }
+
+  Connection(
+      ConnectionContext context,
+      HostInfo hostInfo,
+      String cachedHost,
+      Delegate delegate,
+      String optLastSessionId,
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/database/connection/Connection.java]: Refactored away

File: src/main/java/com/google/firebase/database/connection/PersistentConnectionImpl.java [hunk 12]
```diff
@@ -536,9 +544,9 @@ public void openNetworkConnection(String token) {
     }
     this.authToken = token;
     this.connectionState = ConnectionState.Connecting;
-    realtime =
-        new Connection(this.context, this.hostInfo, this.cachedHost, this, this.lastSessionId);
-    realtime.open();
+    this.realtime = connFactory.newConnection(
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/database/connection/PersistentConnectionImpl.java]: Removed this dereferences from the method body

File: src/test/java/com/google/firebase/database/core/RepoTest.java (hunk 1 above)
Comment (hiranya911) [reply src/test/java/com/google/firebase/database/core/RepoTest.java]: Done

File: src/main/java/com/google/firebase/database/connection/PersistentConnectionImpl.java [hunk 13]
```diff
@@ -1299,4 +1292,27 @@ public RequestResultCallback getOnComplete() {
       return onComplete;
     }
   }
+
+  interface ConnectionFactory {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/database/connection/PersistentConnectionImpl.java]: Are you suggesting we introduce a new generic `ConnectionFactory<T>` type? Looks a little overkill.

File: src/main/java/com/google/firebase/database/connection/Connection.java [hunk 14]
```diff
@@ -262,6 +269,28 @@ public void injectConnectionFailure() {
...
+  interface WebsocketConnectionFactory {
+    WebsocketConnection newConnection(
+        ConnectionContext context,
+        HostInfo hostInfo,
+        String cachedHost,
+        WebsocketConnection.Delegate delegate,
+        String optLastSessionId);
+  }
+
+  private static class DefaultConnectionFactory implements WebsocketConnectionFactory {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/database/connection/Connection.java]: Renamed to `DefaultWebsocketConnectionFactory`

**Comment by hiranya911** [comment]
I've made all the changes except for the use of generics for different factory types. I've left a follow up question regarding that.

File: src/main/java/com/google/firebase/database/connection/PersistentConnectionImpl.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply src/main/java/com/google/firebase/database/connection/PersistentConnectionImpl.java]: I think it made more sense in the original PR as it had a lot of shared method parameters. I like the way it is structured now.

File: src/test/java/com/google/firebase/auth/FirebaseUserManagerTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply src/test/java/com/google/firebase/auth/FirebaseUserManagerTest.java]: Is it possible to flip this map around? You could then use Map builders and remove about 50 lines in this file (No, I didn't count).

[the remark this answers is not in the document]
Comment (hiranya911) [reply src/test/java/com/google/firebase/auth/FirebaseUserManagerTest.java]: Done. (removed 33 lines :) )

**Pull request merged**